File: julia/src/Amplitudes.jl
# File: Amplitudes.jl
# Project: src
# Created Date: 14/06/2022
# Author: Shun Suzuki
# -----
# Last Modified: 08/08/2022
# Modified By: Shun Suzuki (suzuki@hapis.k.u-tokyo.ac.jp)
# -----
# Copyright (c) 2022 Shun Suzuki. All rights reserved.
# 

mutable struct Amplitudes
    _body_ptr::Ptr{Cvoid}
    function Amplitudes(amps::Array{Float64,1})
        chandle = Ref(Ptr{Cvoid}(0))
        autd3capi.autd_create_amplitudes(chandle, amps)
        c = new(chandle[])
        finalizer(c -> autd3capi.autd_delete_amplitudes(c._body_ptr), c)
        c
    end
end

File: matlab/autd3/Amplitudes.m
%{
%File: Amplitudes.m
%Project: autd3
%Created Date: 10/06/2022
%Author: Shun Suzuki
%-----
%Last Modified: 10/06/2022
%Modified By: Shun Suzuki (suzuki@hapis.k.u-tokyo.ac.jp)
%-----
%Copyright (c) 2022 Shun Suzuki. All rights reserved.
%
%}

classdef Amplitudes < Header

    methods

        function obj = Amplitudes(amp)
            obj = obj@Header();
            pp = libpointer('voidPtrPtr', obj.ptr);
            calllib('autd3capi', 'AUTDCreateAmplitudes', pp, amp);
        end

        function delete(obj)

            if obj.ptr.Value ~= 0
                calllib('autd3capi', 'AUTDDeleteAmplitudes', obj.ptr);
                obj.ptr = libpointer('voidPtr', 0);
            end

        end

    end

end
